test/unit/webui/app.spec.js

```diff
@@ -0,0 +1,112 @@
+import React from 'react';
+import { mount } from 'enzyme';
+import { Base64 } from 'js-base64';
+import addHours from 'date-fns/add_hours';
+import storage from '../../../src/webui/utils/storage';
+import App from '../../../src/webui/app';
+
+jest.mock('../../../src/webui/utils/storage', () => {
+  class LocalStorageMock {
+    constructor() {
+      this.store = {};
+    }
+    clear() {
+      this.store = {};
+    }
+    getItem(key) {
+      return this.store[key] || null;
+    }
+    setItem(key, value) {
+      this.store[key] = value.toString();
+    }
+    removeItem(key) {
+      delete this.store[key];
+    }
+  }
+  return new LocalStorageMock();
+});
+
+jest.mock('element-theme-default', () => ({}));
+
+jest.mock('element-react/src/locale/lang/en', () => ({}));
+
+jest.mock('../../../src/webui/utils/api', () => ({
+  request: require('./components/__mocks__/api').default.request
+}));
+
+describe('App', () => {
+  let wrapper;
+
+  beforeEach(() => {
+    wrapper = mount(<App />);
+  });
+  it('loadLogo: set logo url in state', async () => {
+    const { loadLogo } = wrapper.instance();
+    await loadLogo();
+    expect(wrapper.state().logoUrl).toEqual(
+      'http://localhost/-/static/logo.png'
+    );
+  });
+
+  it('toggleLoginModal: should toggle the value in state', () => {
+    const { toggleLoginModal } = wrapper.instance();
+    expect(wrapper.state().showLoginModal).toBeFalsy();
+    toggleLoginModal();
+    expect(wrapper.state('showLoginModal')).toBeTruthy();
+    expect(wrapper.state('error')).toEqual({});
+  });
+
+  it('isUserAlreadyLoggedIn: token already available in storage', async () => {
+    const generateTokenWithTimeRange = (limit = 0) => {
+      const payload = {
+        username: 'verdaccio',
+        exp: Number.parseInt(addHours(new Date(), limit).getTime() / 1000, 10)
```

Comment:
`Number.parseInt(addHours(new Date(), limit).getTime() / 1000, 10)` is duplicated above